IdentityCore/src/cache/metadata/accountMetadata/MSIDAccountMetadataCacheItem.m

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSIDAccountMetadataCacheItem.h"
+#import "MSIDAccountIdentifier.h"
+#import "MSIDAuthority.h"
+#import "MSIDAuthorityFactory.h"
+
+static const NSString *AccountMetadataURLMapKey = @"URLMap";
+
+@implementation MSIDAccountMetadataCacheItem
+
+- (instancetype)initWithHomeAccountId:(NSString *)homeAccountId
+                             clientId:(NSString *)clientId
+
+{
+    if (!homeAccountId || !clientId) return nil;
+    
+    self = [super init];
+    if (self)
+    {
+        _homeAccountId = homeAccountId;
+        _clientId = clientId;
+        _internalMap = [NSMutableDictionary new];
+    }
+    return self;
+}
+
+#pragma mark - URL caching
+- (BOOL)setCachedURL:(NSURL *)cachedURL forRequestURL:(NSURL *)requestURL error:(NSError **)error
+{
+    if ([NSString msidIsStringNilOrBlank:cachedURL.absoluteString]
+        || [NSString msidIsStringNilOrBlank:requestURL.absoluteString])
+    {
+        if (error) *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInvalidInternalParameter, @"Either a target or request URL produces a nil string", nil, nil, nil, nil, nil);
+        
+        return NO;
+    }
+    
+    NSMutableDictionary *urlMap = _internalMap[MSID_ACCOUNT_CACHE_KEY];
+    if (!urlMap)
+    {
+        urlMap = [NSMutableDictionary new];
+        _internalMap[AccountMetadataURLMapKey] = urlMap;
+    }
+    
+    urlMap[requestURL.absoluteString] = cachedURL.absoluteString;
+    return YES;
+}
+
+- (NSURL *)cachedURL:(NSURL *)cachedURL
+{
+    NSDictionary *urlMap = _internalMap[AccountMetadataURLMapKey];
+    NSString *cachedURLString = urlMap[cachedURL.absoluteString];
+    
+    return cachedURLString ? [[NSURL alloc] initWithString:urlMap[cachedURL.absoluteString]] : nil;
```

Comment:
nit: instead of calling "cachedURL.absoluteString", you can use cachedURLString directly